addons/binding/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/handler/SleepIQCloudHandler.java

```diff
@@ -0,0 +1,322 @@
+        }
+
+        logger.debug("Setting SleepIQ cloud online");
+        updateListenerManagement();
+        updateStatus(ThingStatus.ONLINE);
+    }
+
+    /**
+     * Create a new SleepIQ cloud service connection. If a connection already exists, it will be lost.
+     *
+     * @throws LoginException if there is an error while authenticating to the service
+     */
+    private synchronized void createCloudConnection() throws LoginException {
+
+        logger.debug("Reading SleepIQ cloud binding configuration");
+        SleepIQCloudConfiguration bindingConfig = getConfigAs(SleepIQCloudConfiguration.class);
+
+        logger.debug("Creating SleepIQ client");
+        Configuration cloudConfig = new Configuration().withUsername(bindingConfig.username)
+                .withPassword(bindingConfig.password).withLogging(bindingConfig.logging);
+        cloud = SleepIQ.create(cloudConfig);
+
+        logger.info("Authenticating to SleepIQ cloud service");
+        cloud.login();
+
+        logger.info("Authentication successful");
+    }
+
+    @Override
+    public synchronized void dispose() {
+
+        logger.debug("Disposing SleepIQ cloud handler");
+
+        if (pollingJob != null && !pollingJob.isCancelled()) {
+            pollingJob.cancel(true);
+            pollingJob = null;
+        }
+
+        cloud = null;
+    }
+
+    /**
+     * Get the cloud service API.
+     *
+     * @return the service object (may be <code>null</code>)
+     */
+    private synchronized SleepIQ getCloud() {
+        return cloud;
+    }
+
+    /**
+     * Start or stop a background polling job to look for bed status updates based on the whether or not there are any
+     * listeners to notify.
+     */
+    private synchronized void updateListenerManagement() {
+
+        if (cloud == null) {
+            return;
+        }
+
+        if (!bedStatusListeners.isEmpty() && (pollingJob == null || pollingJob.isCancelled())) {
+            int pollingInterval = getConfigAs(SleepIQCloudConfiguration.class).pollingInterval;
+            pollingJob = scheduler.scheduleAtFixedRate(pollingRunnable, pollingInterval, pollingInterval,
+                    TimeUnit.SECONDS);
+        } else if (bedStatusListeners.isEmpty() && pollingJob != null && !pollingJob.isCancelled()) {
+            pollingJob.cancel(true);
+            pollingJob = null;
+        }
+    }
+
+    /**
+     * Retrieve the latest status on all beds and update all registered listeners.
+     */
+    private void publishBedStatusUpdates() {
+        publishBedStatusUpdates(null);
+    }
+
+    /**
+     * Retrieve the latest status on all beds and update the given listener. If the listener is <code>null</code>, all
+     * registered listeners will be updated.
+     *
+     * @param listener the listener to update (may be <code>null</code>)
+     */
+    private void publishBedStatusUpdates(final BedStatusListener listener) {
+
+        SleepIQ cloud = getCloud();
+        if (cloud == null) {
```

Comment:
When I mentioned about references, I was specifically referring to the ```dispose()``` method that changes the internal ```cloud``` state to ```null```. If one thread calls ```dispose()``` while another is executing on the Bridge or any of the Things connected to the Bridge, a NPE will likely occur.

Perhaps it would be better to simply not delete the reference to the cloud object within the ```dispose()``` method and let it get cleaned up when all the references to the Bridge have gone away. That does mean that actions could still occur on the Bridge and its Things after ```dispose()``` is called. Would this approach be preferable?